src/ape_cache/query.py

```diff
@@ -0,0 +1,337 @@
+
+    @perform_query.register
+    def perform_transaction_query(self, query: BlockTransactionQuery) -> pd.DataFrame:
+        with self.engine.connect() as conn:
+            q = conn.execute(
+                text(self.perform_query_stmt(query)),
+                columns=",".join(query.columns),
+                start_block=query.block_id,
+            )
+            return pd.DataFrame(columns=query.columns, data=q.fetchall())
+
+    @perform_query.register
+    def perform_contract_events_query(self, query: ContractEventQuery) -> pd.DataFrame:
+        return self._block_query(self.perform_query_stmt(query), query)  # type: ignore
+
+    @singledispatchmethod
+    def get_dataframe(self, query: QueryType, result: List[Any]):
+        pass
+
+    @get_dataframe.register
+    def get_block_dataframe(self, query: BlockQuery, result: List[Any]):
+        return pd.DataFrame(
+            columns=query.columns,
+            data=[val for val in map(lambda val: val.dict(by_alias=False), result)],
+        )
+
+    @get_dataframe.register
+    def get_transaction_dataframe(self, query: BlockTransactionQuery, result: List[Any]):
+        df = pd.DataFrame(
+            columns=["receiver", "nonce", "sender"],
+            data=[(val.receiver, val.nonce, val.sender) for val in result],
+        )
+        df["block_hash"] = query.block_id
+        if query.columns != ["*"]:
+            df = df[[query.columns]]
+        return df
+
+    @get_dataframe.register
+    def get_contract_event_dataframe(self, query: ContractEventQuery, result: List[Any]):
+        return pd.DataFrame(
+            columns=query.columns,
+            data=[val for val in map(lambda val: val.dict(by_alias=False), result)],
+        )
+
+    def update_cache(self, query: QueryType, result: List[Any]) -> None:
+        df = self.get_dataframe(query, result)
+
+        try:
+            with self.engine.connect() as conn:
+                for idx, row in df.iterrows():
+                    try:
+                        v = conn.execute(
+                            text(self.cache_query(query)),
+                            column=self.column(query),
+                            val=str(row[self.column(query)]),
+                        )
+                        if [i for i in v]:
+                            df = df[df[self.column(query)] != row[self.column(query)]]
+
+                        if df.empty:
+                            return
+
+                    except sqlalchemy.exc.OperationalError as err:
+                        logger.info(err)
+                        df.to_sql(self.table(query), conn, if_exists="append", index=False)
+                        return
+
+                df.to_sql(self.table(query), conn, if_exists="append", index=False)
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.info(err)
```

Comment:
Why is this one at `info` and the rest our at `debug`? Was this supposed to be `debug` like the others?